stdlib/string.gr

```diff
@@ -519,3 +519,596 @@ export let endsWith = (p: String, s: String) => {
     Memory.compare(p, s + n - m, m) == 0n
   }
 }
+
+export enum Encoding {
+  UTF8,
+  UTF16_BE,
+  UTF16_LE,
+  UTF32_BE,
+  UTF32_LE,
+}
+
+// String->Byte encoding and helper functions:
+
+@disableGC
+let utf16Length = (s: String) => {
+  let (>>>) = WasmI32.shrU
+  let (<<) = WasmI32.shl
+  let (+) = WasmI32.add
+  let (&) = WasmI32.and
+  let (<) = WasmI32.ltU
+  let (==) = WasmI32.eq
+
+  let size = WasmI32.fromGrain(byteLength(s)) >>> 1n
+  let len = WasmI32.fromGrain(length(s)) >>> 1n
+
+  let s = WasmI32.fromGrain(s)
+
+  let mut ptr = s + 8n
+  let end = ptr + size
+  let mut size = 0n // <- number of UTF-16 code words
+
+  while (ptr < end) {
+    let byte = WasmI32.load8U(ptr, 0n)
+    let n = if ((byte & 0x80n) == 0x00n) {
+      1n
+    } else if ((byte & 0xF0n) == 0xF0n) {
+      4n
+    } else if ((byte & 0xE0n) == 0xE0n) {
+      3n
+    } else {
+      2n
+    }
+    if (n == 4n) {
+      size += 2n
+    } else {
+      size += 1n
+    }
+    ptr += n
+  }
+  // multiply by two to get number of bytes (equivalent to left-shift)
```

Comment:
I read this and expected to see multiplication instead of a shift, and then I saw a shift and it made me think that the code was originally a multiplication and then it was changed. Regardless I think I wouldn't have batted an eye if it didn't have this last part 😂 
```suggestion
  // multiply by two to get number of bytes
```